lib/mutgen.c

```diff
@@ -433,28 +710,289 @@ mutgen_generate(mutgen_t *self, tsk_table_collection_t *tables, int flags)
+            site_left = discrete_sites ? ceil(left) : left;
+            site_right = discrete_sites ? ceil(right) : right;
+            mu = branch_length * (site_right - site_left) * map_rate[map_index];
             branch_mutations = gsl_ran_poisson(self->rng, mu);
             for (l = 0; l < branch_mutations; l++) {
-                /* Rejection sample positions until we get one we haven't seen before. */
-                /* TODO add a maximum number of rejections here */
-                do {
-                    position = gsl_ran_flat(self->rng, left, right);
+                /* Rejection sample positions until we get one we haven't seen before,
+                 * unless we are doing discrete sites. */
+                for (k = 0; k < max_num_rejections; k++) {
+                    position = gsl_ran_flat(self->rng, site_left, site_right);
+                    if (discrete_sites)
+                        position = floor(position);
                     search.position = position;
                     avl_node = avl_search(&self->sites, &search);
-                } while (avl_node != NULL);
-                assert(left <= position && position < right);
-                type = gsl_rng_uniform_int(self->rng, num_mutation_types);
-                ret = mutgen_add_mutation(self, child, position,
-                        mutation_types[type].ancestral_state,
-                        mutation_types[type].derived_state);
+                    if (avl_node == NULL)
+                        break;
+                }
+                /* TODO: what happens if max_num_rejections is reached
+                 * and we're not in discrete land? */
+                time = gsl_ran_flat(self->rng, branch_start, branch_end);
+                assert(site_left <= position && position < site_right);
+                assert(branch_start <= time && time < branch_end);
+                if (avl_node == NULL) {
+                    ret = mutgen_add_site(self, position, true,
+                            &dummy_alphabet[0], 1,
+                            NULL, 0, &avl_node);
+                    if (ret != 0) {
+                        goto out;
+                    }
+                }
+                ret = mutgen_add_mutation(self, TSK_NULL, child, time, true,
+                        &dummy_alphabet[1], 1,
+                        NULL, 0, avl_node);
                 if (ret != 0) {
                     goto out;
                 }
             }
             map_index++;
         }
     }
+out:
+    return ret;
+}
+
+static int
+mutgen_pick_allele(mutgen_t *self)
+{
+    mutation_model_t *model = self->model;
+    double u = gsl_ran_flat(self->rng, 0.0, 1.0);
+    int j = 0;
+    while (u > model->root_distribution[j]) {
+        u -= model->root_distribution[j];
+        j++;
+        assert(j < model->alphabet_length);
+    }
+    return j;
+}
+
+static int
+mutgen_transition_allele(mutgen_t *self, int parent_allele)
```

Comment:
This is a method of the mutation_model_t class, isn't it?

Reply:
It would be except that it needs `rng`. We could pass in `rng`, but this seems fine?

Reply:
Sure, passing in rng is perfectly reasonable.